src/lib/crypto/kimchi_backend/gadgets/foreign_field.ml

```diff
@@ -106,188 +121,660 @@ let field_standard_limbs_to_bignum_bigint (type f)
+        (* Return the ffadd values *)
+        [| result0; result1; result2; field_overflow; carry_bot |] )
+    |> tuple5_of_array
+  in
 
-  let of_bignum_bigint (type field)
-      (module Circuit : Snark_intf.Run with type field = field) x : field t =
-    let open Circuit in
-    let l12, l0 = Common.(bignum_bigint_div_rem x two_to_limb) in
-    let l2, l1 = Common.(bignum_bigint_div_rem l12 two_to_limb) in
-    let l0 =
-      Field.constant @@ Common.bignum_bigint_to_field (module Circuit) l0
-    in
-    let l1 =
-      Field.constant @@ Common.bignum_bigint_to_field (module Circuit) l1
-    in
-    let l2 =
-      Field.constant @@ Common.bignum_bigint_to_field (module Circuit) l2
+  (* Create the gate *)
+  with_label "ffadd_gate" (fun () ->
+      (* Set up FFAdd gate *)
+      assert_
+        { annotation = Some __LOC__
+        ; basic =
+            Kimchi_backend_common.Plonk_constraint_system.Plonk_constraint.T
+              (ForeignFieldAdd
+                 { left_input_lo = left_input0
+                 ; left_input_mi = left_input1
+                 ; left_input_hi = left_input2
+                 ; right_input_lo = right_input0
+                 ; right_input_mi = right_input1
+                 ; right_input_hi = right_input2
+                 ; field_overflow
+                 ; carry
+                 ; foreign_field_modulus0
+                 ; foreign_field_modulus1
+                 ; foreign_field_modulus2
+                 ; sign
+                 } )
+        } ) ;
+
+  (* Return the result *)
+  (Element.Standard.of_limbs (result0, result1, result2), sign, field_overflow)
+
+(* Gadget to check the supplied value is a valid foreign field element for the
+ * supplied foreign field modulus
+ *
+ *    This gadget checks in the circuit that a value is less than the foreign field modulus.
+ *    Part of this involves computing a bound value that is both added to external_checks
+ *    and also returned.  The caller may use either one, depending on the situation.
+ *
+ *    Inputs:
+ *      external_checks       := Context to track required external checks
+ *      value                 := the value to check
+ *      foreign_field_modulus := the modulus of the foreign field
+ *
+ *    Outputs:
+ *      Inserts the gates (described below) into the circuit
+ *      Adds bound value to be multi-range-checked to external_checks
+ *      Returns bound value
+ *
+ *    Effects to the circuit:
+ *      - 1 FFAdd gate
+ *      - 1 Zero gate
+ *)
+let valid_element (type f) (module Circuit : Snark_intf.Run with type field = f)
+    (external_checks : f External_checks.t) (value : f Element.Standard.t)
+    (foreign_field_modulus : f standard_limbs) : f Element.Standard.t =
+  let open Circuit in
+  (* Compute the value for the right input of the addition as 2^264 *)
+  let offset0 = Field.zero in
+  let offset1 = Field.zero in
+  let offset2 =
+    exists Field.typ ~compute:(fun () -> two_to_limb_field (module Circuit))
```

Comment:
Mmmm.... Not sure I see the problem... If you look at line 586, there is a `Field.Assert.equal` that forces it to be exactly 2^88, so setting it to 0 wouldn't work:
```
  (* Check that the highest limb of right input is 2^88*)
  let two_to_88 = two_to_limb_field (module Circuit) in
  Field.Assert.equal (Field.constant two_to_88) offset2 ;
```
If you're wondering why I am not asserting it right after instead of a few lines later, that's just to avoid the possibility that a generic gate appears in the middle of the gadget, perhaps breaking the curr-next row consistency. Thus, I wait until the Zero gate is added, to not only constrain that `right2` must be 2^88 (so that right is 2^264), but also that the overflow bit must be one. 

Does that address the issue?

Reply:
Aha, I didn't mentally tie the knot on that one. Awesome!